src/computability/reduce.lean

```diff
@@ -258,108 +255,165 @@ theorem disjoin_le {α β γ} [primcodable α] [primcodable β] [primcodable γ]
   one_one_reducible.disjoin_right.to_many_one.trans h⟩,
  λ ⟨h₁, h₂⟩, disjoin_many_one_reducible h₁ h₂⟩
 
+/-- A many-one degree is an equivalence class of sets up to many-one equivalence. -/
+def many_one_degree : Type :=
+quotient $ show setoid (Σ (A : set ℕ) [primcodable A], set A), from
+setoid.mk (λ ⟨_, _, p⟩ ⟨_, _, q⟩, by exactI many_one_equiv p q)
+  ⟨λ ⟨_, _, a⟩, by exactI many_one_equiv_refl _,
+   λ ⟨_, _, _⟩ ⟨_, _, b⟩ h, by exactI h.symm,
+   λ ⟨_, _, a⟩ ⟨_, _, b⟩ ⟨_, _, c⟩ hab hbc, by exactI hab.trans hbc⟩
+
+namespace many_one_degree
+variables {α : Type u} [primcodable α]
+variables {β : Type v} [primcodable β]
+variables {γ : Type w} [primcodable γ]
+
+/-- The many-one degree of a set on a primcodable type. -/
+def of (p : α → Prop) : many_one_degree :=
+quotient.mk' ⟨_, primcodable.ulower, p ∘ ulower.up⟩
+
+@[elab_as_eliminator]
+protected lemma ind_on {C : many_one_degree → Prop} (d : many_one_degree)
+  (h : ∀ (α : Type) [primcodable α] (p : α → Prop), C (of p)) : C d :=
```

Comment:
Why does this induction principle quantify over all primcodable types? Shouldn't `nat` suffice?

Reply:
You never need the fact that `α` is actually a subset of the natural numbers.  IMHO it would just make the goal more messy,